runtime/compiler/runtime/JITServerAOTDeserializer.cpp

```diff
@@ -898,3 +901,600 @@ JITServerLocalSCCAOTDeserializer::updateSCCOffsets(SerializedAOTMethod *method,
+   {
+   OMR::CriticalSection cs(getClassLoaderMonitor());
+   if (deserializerWasReset(comp, wasReset))
+      return false;
+
+   auto it = _classLoaderIdMap.find(record->id());
+   if (it != _classLoaderIdMap.end())
+      return true;
+   isNew = true;
+
+   // Lookup the class loader using the name of the first class that it loaded
+   auto loader = (J9ClassLoader *)getLoaderTable()->lookupClassLoaderAssociatedWithClassName(record->name(), record->nameLength());
+   if (!loader)
+      {
+      if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+         TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+            "ERROR: Failed to find class loader for first loaded class %.*s", RECORD_NAME(record)
+         );
+      return false;
+      }
+
+   addToMaps(_classLoaderIdMap, _classLoaderPtrMap, it, record->id(), loader, loader);
+
+   if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+         "Cached class loader record ID %zu -> { %p } for first loaded class %.*s",
+         record->id(), loader, RECORD_NAME(record)
+      );
+   return true;
+   }
+
+bool
+JITServerNoSCCAOTDeserializer::cacheRecord(const ClassSerializationRecord *record, TR::Compilation *comp, bool &isNew, bool &wasReset)
+   {
+   OMR::CriticalSection cs(getClassMonitor());
+   if (deserializerWasReset(comp, wasReset))
+      return false;
+
+   auto it = _classIdMap.find(record->id());
+   if (it != _classIdMap.end())
+      {
+      if (it->second._ramClass)
+         {
+         return true;
+         }
+      else
+         {
+         if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+            TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "ERROR: Mismatching class ID %zu", record->id());
+         return false;
+         }
+      }
+   isNew = true;
+
+   // The class loader for this class record should already have been deserialized, so if we can't find a
+   // loader for this ID then it must have been marked as unloaded. We don't support loader reloading, so
+   // we simply fail to deserialize here.
+   auto loader = findInMap(_classLoaderIdMap, record->classLoaderId(), getClassLoaderMonitor(), comp, wasReset);
+   if (!loader)
+      return false;
+
+   // Lookup the RAMClass by name in the class loader
+   J9Class *ramClass = jitGetClassInClassloaderFromUTF8(comp->j9VMThread(), loader, (char *)record->name(),
+                                                        record->nameLength());
+   if (!ramClass)
+      {
+      if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+         TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+            "ERROR: Failed to find class %.*s ID %zu in class loader %p",
+            RECORD_NAME(record), record->id(), loader
+         );
+      return false;
+      }
+
+   // Check that the ROMClass hash matches, otherwise remember that it doesn't
```

Comment:
Why remember it? Is this info used somewhere?

Reply:
The info that it doesn't match? It's how the original deserializer functions. I think it exists to avoid looking up and re-hashing the same mismatching class repeatedly. Also it may interact with the new known IDs work - we still mark this record as new so the server will stop sending it to the client, so we still need to keep an entry in the map that indicates that there was a problem with deserializing that one particular record originally.